File: 2025/day07/cpp/day07.cpp
#include <fstream>
#include <iostream>
#include <string>
#include <filesystem>
#include <vector>
#include <unordered_set>
#include <unordered_map>
#include "default_grid.h"
#include "utils.h"

using namespace std;

using ll = long long;
using range = pair<ll, ll>;
using Grid = DefaultGrid<char, '.'>;

unordered_set<pair<int, int>, IntPairPackerFunctor> splitters_reached;
unordered_map<pair<int, int>, ll, IntPairPackerFunctor> memo;

void parse(const string& filename, Grid& grid, pair<int, int>& s_pos)
{
	if (!filesystem::exists(filename))
	{
		string fileInfo;
		fileInfo += "This is the current working directory: " + filesystem::current_path().string() + ".\n";
		fileInfo += "Filename supplied: " + filename + ".\n";
		fileInfo += "Resolves to absolute path: " + filesystem::absolute(filename).string() + ".\n";

		cerr << fileInfo << endl;
		cerr << "Error: file does not exist: " << filesystem::absolute(filename) << endl;
		throw runtime_error("file does not exist");
	}

	ifstream f(filename);
	if (!f.is_open())
	{
		cerr << "Error opening file: " << filename << endl;
		throw runtime_error("error opening file");
	}

	string line;
	int row_idx = 0;
	int col_idx = 0;
	while (getline(f, line))
	{
		col_idx = 0;
		for (char c : line)
		{
			grid.set_at(row_idx, col_idx, c);

			if (c == 'S')
			{
				s_pos = {row_idx, col_idx};
			}
			col_idx++;
		}

		row_idx++;
	}
}

void beam_from_point_p1(Grid& grid, int start_row, int start_col)
{
	if (start_row >= grid.rows() || start_col < 0 || start_col >= grid.cols())
		return;

	char curr_val = grid.get_at(start_row, start_col);

	if (curr_val == '.')
	{
		grid.set_at(start_row, start_col, '|');
		beam_from_point_p1(grid, start_row + 1, start_col);
	}
	else if (curr_val == '^')
	{
		splitters_reached.insert({start_row, start_col});
		beam_from_point_p1(grid, start_row, start_col - 1);
		beam_from_point_p1(grid, start_row, start_col + 1);
	}
}

ll beam_from_point_p2(Grid& grid, int start_row, int start_col)
{
	if (start_row >= grid.rows() || start_col < 0 || start_col >= grid.cols())
		return 1;

	if (memo.contains({start_row, start_col}))
		return memo[{start_row, start_col}];

	char curr_val = grid.get_at(start_row, start_col);

	ll timelines = 0;
	if (curr_val == '.')
	{
		timelines += beam_from_point_p2(grid, start_row + 1, start_col);
	}
	else if (curr_val == '^')
	{
		timelines += beam_from_point_p2(grid, start_row, start_col - 1);
		timelines += beam_from_point_p2(grid, start_row, start_col + 1);
	}

	memo[{start_row, start_col}] = timelines;
	return timelines;
}

void part1(Grid grid, const pair<int, int>& s_pos)
{
	splitters_reached.clear();
	
	int s_col = s_pos.second;
	beam_from_point_p1(grid, 1, s_col);

	auto ans = splitters_reached.size();
	cout << "Part 1: " << ans << endl;
}

void part2(Grid grid, const pair<int, int>& s_pos)
{
	memo.clear();
	
	int s_col = s_pos.second;
	auto timelines = beam_from_point_p2(grid, 1, s_col);

	cout << "Part 2: " << timelines << endl;
}

int main()
{
	//string filename = "../../../../example";
	string filename = "../../../../input";

	Grid grid;
	pair<int, int> s_pos;
	parse(filename, grid, s_pos);

	part1(Grid(grid), s_pos); // copy constructor so we don't modify original grid
	part2(grid, s_pos);

	return 0;
}

File: 2025/day07/py/day07.py
import sys
import os

# silly python path manipulation
sys.path.append(os.path.abspath(os.path.join(os.path.dirname(__file__), "../..")))

from utils.grid import Grid

splitters_reached = set()
memo = {}


def parse(filename: str) -> Grid:
    with open(filename) as f:
        lines = [line.strip() for line in f.readlines()]

    grid = Grid(len(lines), len(lines[0]), default_value=".")
    for row_idx, line in enumerate(lines):
        for col_idx, ch in enumerate(line):
            if ch != ".":
                grid.set(row_idx, col_idx, ch)

    return grid


def beam_from_point_p1(grid: Grid, start_row: int, start_col: int):
    if start_row >= grid.rows or start_col < 0 or start_col >= grid.cols:
        return

    curr_val = grid.get(start_row, start_col)

    if curr_val == ".":
        grid.set(start_row, start_col, "|")
        beam_from_point_p1(grid, start_row + 1, start_col)
    elif curr_val == "^":
        splitters_reached.add((start_row, start_col))
        beam_from_point_p1(grid, start_row, start_col - 1)
        beam_from_point_p1(grid, start_row, start_col + 1)

    return


def beam_from_point_p2(grid: Grid, start_row: int, start_col: int) -> int:
    if start_row >= grid.rows or start_col < 0 or start_col >= grid.cols:
        return 1

    if (start_row, start_col) in memo:
        return memo[(start_row, start_col)]

    curr_val = grid.get(start_row, start_col)

    timelines = 0
    if curr_val == ".":
        timelines += beam_from_point_p2(grid, start_row + 1, start_col)
    elif curr_val == "^":
        timelines += beam_from_point_p2(grid, start_row, start_col - 1)
        timelines += beam_from_point_p2(grid, start_row, start_col + 1)

    memo[(start_row, start_col)] = timelines
    return timelines


def find_s(grid: Grid):
    s_col = -1
    for col_idx in range(grid.cols):
        val = grid.get(0, col_idx)
        if val == "S":
            s_col = col_idx
            break

    assert s_col != -1
    return s_col


def part1(grid: Grid):
    s_col = find_s(grid)
    beam_from_point_p1(grid, 1, s_col)

    ans = len(splitters_reached)

    print("Part 1:", ans)


def part2(grid: Grid):
    s_col = find_s(grid)
    timelines = beam_from_point_p2(grid, 1, s_col)

    print("Part 2:", timelines)


# filename = "day07/example"
filename = "day07/input"

grid = parse(filename)
part1(grid.copy())
part2(grid.copy())
